src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/BinaryEncoders/CompressedMessageBinaryEncoder.cs

```diff
@@ -0,0 +1,226 @@
+﻿/* Copyright 2013-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver.Core.Compression;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.WireProtocol.Messages.Encoders.BinaryEncoders
+{
+    /// <summary>
+    /// Represents a binary encoder for a compressed message.
+    /// </summary>
+    public sealed class CompressedMessageBinaryEncoder : MessageBinaryEncoderBase, IMessageEncoder
+    {
+        private readonly ICompressorSource _compressorSource;
+        private readonly MessageEncoderSettings _encoderSettings;
+        private const int MessageHeaderLength = 16;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessageBinaryEncoder" /> class.
+        /// </summary>
+        /// <param name="stream">The stream.</param>
+        /// <param name="compressorSource">The compressor source.</param>
+        /// <param name="encoderSettings">The encoder settings.</param>
+        public CompressedMessageBinaryEncoder(
+            Stream stream,
+            ICompressorSource compressorSource,
+            MessageEncoderSettings encoderSettings)
+            : base(stream, encoderSettings)
+        {
+            _compressorSource = Ensure.IsNotNull(compressorSource, nameof(compressorSource));
+            _encoderSettings = encoderSettings; // can be null
+        }
+
+        /// <summary>
+        /// Gets a flag whether the message compressed or not.
+        /// </summary>
+        /// <returns>A compressed flag.</returns>
+        public bool IsCompressed()
+        {
+            var stream = CreateBinaryReader().BsonStream;
+
+            var position = stream.Position;
+
+            // Skip header to opcode
+            stream.ReadInt32();
+            stream.ReadInt32();
+            stream.ReadInt32();
+            var isCompressed = stream.ReadInt32() == (int)Opcode.Compressed;
+
+            stream.Position = position;
+
+            return isCompressed;
+        }
+
+        /// <summary>
+        /// Reads the message.
+        /// </summary>
+        /// <returns>A message.</returns>
+        public CompressedMessage ReadMessage()
```

Comment:
I propose this implementation instead:

        public CompressedMessage ReadMessage()
        {
            var reader = CreateBinaryReader();
            var stream = reader.BsonStream;

            var messageStartPosition = stream.Position;
            var messageLength = stream.ReadInt32();
            EnsureMessageLengthIsValid(messageLength);
            var requestId = stream.ReadInt32();
            var responseTo = stream.ReadInt32();
            var opcode = (Opcode)stream.ReadInt32();
            EnsureOpcodeIsValid(opcode);
            var originalOpcode = (Opcode)stream.ReadInt32();
            var uncompressedSize = stream.ReadInt32();
            var compressorType = (CompressorType)stream.ReadByte();
            var compressor = _compressorSource.Get(compressorType);

            using (var uncompressedBuffer = new MultiChunkBuffer(new OutputBufferChunkSource(BsonChunkPool.Default)))
            using (var uncompressedStream = new ByteBufferStream(uncompressedBuffer, ownsBuffer: false))
            {
                uncompressedStream.WriteInt32(uncompressedSize + MessageHeaderLength);
                uncompressedStream.WriteInt32(requestId);
                uncompressedStream.WriteInt32(responseTo);
                uncompressedStream.WriteInt32((int)originalOpcode);
                compressor.Decompress(stream, uncompressedStream);
                uncompressedStream.Position = 0;

                var originalMessageEncoderFactory = new BinaryMessageEncoderFactory(uncompressedStream, _encoderSettings, _compressorSource);
                var originalMessageEncoder = _originalEncoderSelector.GetEncoder(originalMessageEncoderFactory);
                var originalMessage = originalMessageEncoder.ReadMessage();

                return new CompressedMessage(originalMessage);
            }
        }